chapter_recurrent-neural-networks/text-preprocessing.md

```diff
@@ -72,113 +66,88 @@ print(lines[10])
-    elif token == 'char':
-        return [list(line) for line in lines]
-    else:
-        print('ERROR: unknown token type: ' + token)
+    assert token in ('word', 'char'), 'unknown token type: ' + token
+    return [line.split() if token == 'word' else list(line) for line in lines]
 
 tokens = tokenize(lines)
-for i in range(11):
-    print(tokens[i])
+for i in range(7, 10):
+    print(f'line {i+1}: {tokens[i]}')
 ```
 
 ## Vocabulary
 
-The string type of the token is inconvenient to be used by models, which take numerical inputs.
-Now let us [**build a dictionary, often called *vocabulary* as well, to map string tokens into numerical indices starting from 0**].
-To do so, we first count the unique tokens in all the documents from the training set,
-namely a *corpus*,
-and then assign a numerical index to each unique token according to its frequency.
-Rarely appeared tokens are often removed to reduce the complexity.
-Any token that does not exist in the corpus or has been removed is mapped into a special unknown token “&lt;unk&gt;”.
-We optionally add a list of reserved tokens, such as
-“&lt;pad&gt;” for padding,
-“&lt;bos&gt;” to present the beginning for a sequence, and “&lt;eos&gt;” for the end of a sequence.
+While these tokens are still strings,
+our models require numerical inputs.
+[**To this end, we will need a class
+to construct a *vocabulary*
+that assigns a unique index 
+to each distinct token value.**]
+First, we count the occurrences 
+of each element of the vocabulary,
+lumping the rarest ones all together
+into a special value "&lt;unk&gt;" (unknown token).
+In the future,
+we may supplement the vocabulary
+with a list of reserved tokens.
 
 ```{.python .input}
 #@tab all
 class Vocab:  #@save
     """Vocabulary for text."""
-    def __init__(self, tokens=None, min_freq=0, reserved_tokens=None):
-        if tokens is None:
-            tokens = []
-        if reserved_tokens is None:
-            reserved_tokens = []
-        # Sort according to frequencies
-        counter = count_corpus(tokens)
-        self._token_freqs = sorted(counter.items(), key=lambda x: x[1],
-                                   reverse=True)
-        # The index for the unknown token is 0
-        self.idx_to_token = ['<unk>'] + reserved_tokens
+    def __init__(self, tokens=[], min_freq=0, reserved_tokens=[]):
+        # Expand a list of list into a list if needed
+        if tokens and isinstance(tokens[0], list):
+            tokens = [token for line in tokens for token in line]
+        # Count token frequencies
+        counter = collections.Counter(tokens)
+        self.token_freqs = sorted(counter.items(), key=lambda x: x[1],
+                                  reverse=True)
+        # The list of unique tokens. Sorting it for a deterministic mapping
+        self.idx_to_token = list(sorted(set(['<unk>'] + reserved_tokens + [
+            token for token, freq in self.token_freqs if freq >= min_freq])))
```

Comment:
`token_freqs` is already sorted so we can get a deterministic mapping from it without another sorting. Besides, it's quite common to place <unk> at index 0, followed by reserved tokens such as <pad>, <cls>, ...